File: day08/2fact.py
#!/usr/bin/env python
from sys import argv
import re

def init_screen(rows, cols) :
    screen = [[' ' for x in range(cols)] for y in range(rows)]
    return screen

def print_screen(screen, rows, cols) :
    for row in range(rows) :
        for col in range(cols) :
            print screen[row][col],
        print
    print

def do_rect(screen, tok) :
    wh = tok[1].split('x')
    w = int(wh[0])
    h = int(wh[1])
    for row in range(h) :
        for col in range(w) :
            screen[row][col] = '#'

def do_rotate_col(screen, tok, rows) :
    tok2 = tok[2].split('=')
    col = int(tok2[1])
    n = int(tok[4])

    for i in range(1,n+1) :
        tmp = screen[rows-1][col]
        for row in range(rows-1, 0, -1) :
            screen[row][col] = screen[row-1][col]
        screen[0][col] = tmp

def do_rotate_row(screen, tok, cols) :
    tok2 = tok[2].split('=')
    row = int(tok2[1])
    n = int(tok[4])

    for i in range(1, n+1) :
        tmp = screen[row][cols-1]
        for col in range(cols-1, 0, -1) :
            screen[row][col] = screen[row][col-1]
        screen[row][0] = tmp

def num_lit(screen, rows, cols) :
    n = 0

    for row in range(rows) :
        for col in range(cols) :
            if screen[row][col] == '#' :
                n = n + 1

    return n

#ROWS = 3
#COLS = 7
ROWS = 6
COLS = 50
screen = init_screen(ROWS, COLS)
print_screen(screen, ROWS, COLS)

script, filename = argv
with open(filename) as f :
    for line in f :
        line = line.rstrip("\n")
        print line
        tok = line.split(' ')
        if tok[0] == "rect" :
            do_rect(screen, tok)
        elif tok[0] == 'rotate' :
            if tok[1] == 'column' :
                do_rotate_col(screen, tok, ROWS)
            else :
                do_rotate_row(screen, tok, COLS)
                
        print_screen(screen, ROWS, COLS)
    
print num_lit(screen, ROWS, COLS)

File: day08/2fact.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.24.0;
use feature 'signatures';
no warnings "experimental::signatures";

# my $ROWS = 3;
# my $COLS = 7;
my $ROWS = 6;
my $COLS = 50;
my $screen = init($ROWS, $COLS);
print_screen($screen, $ROWS, $COLS);

while (<>) {
    print;
    chomp;
    my @tok = split ' ';
    if ($tok[0] eq 'rect') {
        do_rect($screen, \@tok);
    } elsif ($tok[0] eq 'rotate') {
        if ($tok[1] eq 'column') {
            do_rotate_col($screen, \@tok, $ROWS);
        } else {
            do_rotate_row($screen, \@tok, $COLS);
        }
    }
    print_screen($screen, $ROWS, $COLS);

}

say num_lit($screen, $ROWS, $COLS);

sub init($rows, $cols) {
    my @screen;

    push @screen, [map { ' ' } 1..$cols] for 1..$rows;

    return \@screen;
}

sub print_screen($screen, $rows, $cols) {
    for my $row (0..$rows-1) {
        for my $col (0..$cols-1) {
            print $screen->[$row][$col];
        }
        print "\n";
    }
    print "\n";
}

sub do_rect($screen, $tok) {
    my ($w, $h) = split 'x', $tok->[1];

    for my $row (0..$h-1) {
        for my $col (0..$w-1) {
            $screen->[$row][$col] = '#';
        }
    }
}

sub do_rotate_col($screen, $tok, $rows) {
    my (undef, $col) = split '=', $tok->[2];
    my $n = $tok->[4];

    for my $i (1..$n) {
        my $tmp = $screen->[$rows-1][$col];
        for (my $row = $rows-1; $row >= 1; $row--) {
            $screen->[$row][$col] = $screen->[$row-1][$col];
        }
        $screen->[0][$col] = $tmp;
    }
}

sub do_rotate_row($screen, $tok, $cols) {
    my (undef, $row) = split '=', $tok->[2];
    my $n = $tok->[4];

    for my $i (1..$n) {
        my $tmp = $screen->[$row][$cols-1];
        for (my $col = $cols-1; $col >= 1; $col--) {
            $screen->[$row][$col] = $screen->[$row][$col-1];
        }
        $screen->[$row][0] = $tmp;
    }
}

sub num_lit($screen, $rows, $cols) {
    my $n = 0;

    for my $row (0..$rows-1) {
        for my $col (0..$cols-1) {
            $n++ if $screen->[$row][$col] eq '#';
        }
    }

    return $n;
}
